pyiceberg/table/__init__.py

```diff
@@ -340,6 +341,86 @@ def set_properties(self, properties: Properties = EMPTY_DICT, **kwargs: Any) ->
         updates = properties or kwargs
         return self._apply((SetPropertiesUpdate(updates=updates),))
 
+    @deprecated(
+        deprecated_in="0.7.0",
+        removed_in="0.8.0",
+        help_message="Please use one of the functions in ManageSnapshots instead",
+    )
+    def add_snapshot(self, snapshot: Snapshot) -> Transaction:
+        """Add a new snapshot to the table.
+
+        Returns:
+            The transaction with the add-snapshot staged.
+        """
+        updates = (AddSnapshotUpdate(snapshot=snapshot),)
+        requirements = (AssertTableUUID(uuid=self._table.metadata.table_uuid),)
+
+        return self._apply(updates, requirements)
+
+    @deprecated(
+        deprecated_in="0.7.0",
+        removed_in="0.8.0",
+        help_message="Please use one of the functions in ManageSnapshots instead",
+    )
+    def set_ref_snapshot(
+        self,
+        snapshot_id: int,
+        parent_snapshot_id: Optional[int],
+        ref_name: str,
+        type: str,
+        max_ref_age_ms: Optional[int] = None,
+        max_snapshot_age_ms: Optional[int] = None,
+        min_snapshots_to_keep: Optional[int] = None,
+    ) -> Transaction:
+        """Update a ref to a snapshot.
+
+        Returns:
+            The transaction with the set-snapshot-ref staged
+        """
+        updates = (
+            SetSnapshotRefUpdate(
+                snapshot_id=snapshot_id,
+                ref_name=ref_name,
+                type=type,
+                max_ref_age_ms=max_ref_age_ms,
+                max_snapshot_age_ms=max_snapshot_age_ms,
+                min_snapshots_to_keep=min_snapshots_to_keep,
+            ),
+        )
+
+        requirements = (
+            AssertRefSnapshotId(snapshot_id=parent_snapshot_id, ref="main"),
+            AssertTableUUID(uuid=self.table_metadata.table_uuid),
+        )
+        return self._apply(updates, requirements)
+
+    def _set_ref_snapshot(
+        self,
+        snapshot_id: int,
+        ref_name: str,
+        type: str,
+        max_ref_age_ms: Optional[int] = None,
+        max_snapshot_age_ms: Optional[int] = None,
+        min_snapshots_to_keep: Optional[int] = None,
+    ) -> UpdatesAndRequirements:
+        """Update a ref to a snapshot.
+
+        Returns:
+            The updates and requirements for the set-snapshot-ref staged
+        """
+        updates = (
+            SetSnapshotRefUpdate(
+                snapshot_id=snapshot_id,
+                ref_name=ref_name,
+                type=type,
+                max_ref_age_ms=max_ref_age_ms,
+                max_snapshot_age_ms=max_snapshot_age_ms,
+                min_snapshots_to_keep=min_snapshots_to_keep,
+            ),
+        )
+
+        return updates, ()
```

Comment:
I think we want to require here as well that the ref didn't exist, or that the ref is at a certain snapshot.

Reply:
Something like:
```suggestion
        return updates, (
            AssertRefSnapshotId(
                snapshot_id=snapshot_id,
                # None means that the branch/tag should not exist
                ref=self.table_metadata.refs[ref_name].snapshot_id if ref_name in self.table_metadata.refs else None
            )
        )
```

Reply:
should i still make this change? 

Also, If we get our ref as above, I think it would make the check redundant.
```
# None means that the branch/tag should not exist
ref=self.table_metadata.refs[ref_name] if ref_name in self.table_metadata.refs else None
```
AssertRef is validating the ref from table.metadata.refs with the ref we provide in the function, to make sure that no concurrent writers have changed the ref during the commit.

Reply:
> AssertRef is validating the ref from table.metadata.refs with the ref we provide in the function, to make sure that no concurrent writers have changed the ref during the commit.

This is exactly the thing that we try to accomplish:

- We fetch the table
- Stage an update of the ref and require certain conditions of the current refs
- Try to acquire the lock (in case of Hive)
  - Another process releases the lock, and also updates the same ref
- We validate our `AssertRefSnapshotId`:
  - We observe that the ref is already created or the snapshot-id is different than we expected. In this case we raise an error.
  - All the requirements still hold, and we can update the table.
- We release the lock